File: mancala.c
#include "mancala.h"

#include <stdlib.h>

struct mancala *mancala_new(void)
{
    static const struct mancala starting_state = {
        0,
        {{4, 4, 4, 4, 4, 4}, 0},
        {{4, 4, 4, 4, 4, 4}, 0}};

    struct mancala *game = malloc(sizeof(*game));

    if (game != NULL)
        *game = starting_state;

    return game;
}

static int mancala_do_turn_in_place(struct mancala *game, int index)
{
    unsigned short x;
    struct player *near = game->turn == 0 ? &game->p1 : &game->p2,
                  *far = game->turn == 0 ? &game->p2 : &game->p1;

    if (index >= 0 && index < NUM_CUPS && (x = near->cups[index]) > 0)
    {
        int i;
        unsigned short q = x / (NUM_CUPS * 2 + 1),
                       r = x % (NUM_CUPS * 2 + 1);

        /* distribute stones to cups */
        near->cups[index] = 0;

        for (i = 0; i < NUM_CUPS; i++)
        {
            int dist = i > index
                           ? i - index - 1
                           : i - index + NUM_CUPS * 2;

            near->cups[i] += q + (dist < r);
        }

        for (i = 0; i < NUM_CUPS; i++)
            far->cups[i] += q + (i - index + NUM_CUPS < r);

        near->mancala += q + ((NUM_CUPS - index - 1) < r);

        /* opposite cup stealing */
        if (index + r >= 0 && index + r < NUM_CUPS)
        {
            unsigned short
                *last_cup = &near->cups[index + r],
                *opposite = &far->cups[NUM_CUPS - (index + r) - 1];

            if (*last_cup == 1 && *opposite > 0)
            {
                near->mancala += 1 + *opposite;
                *last_cup = *opposite = 0;
            }
        }

        /* go again if last stone dropped in mancala */
        if (index + r != NUM_CUPS)
        {
            game->turn = !game->turn;
        }

        return 1;
    }

    return 0;
}

struct mancala *mancala_do_turn(struct mancala const *game, int index)
{
    struct player const *near = game->turn == 0 ? &game->p1 : &game->p2;

    if (index >= 0 && index < NUM_CUPS && near->cups[index] > 0)
    {
        struct mancala *copy = malloc(sizeof(*copy));

        if (copy != NULL)
        {
            *copy = *game;
            mancala_do_turn_in_place(copy, index);
        }

        return copy;
    }
    else
        return NULL;
}

static int all_cups_empty(struct player const *player)
{
    int i = 0;

    for (i = 0; i < NUM_CUPS; i++)
    {
        if (player->cups[i] != 0)
            return 0;
    }
    return 1;
}

int mancala_game_over(struct mancala const *game)
{
    return all_cups_empty(&game->p1) || all_cups_empty(&game->p2);
}

void mancala_print(struct mancala const *game, FILE *out)
{
    int i;
    struct player const *near = game->turn == 0 ? &game->p1 : &game->p2,
                        *far = game->turn == 0 ? &game->p2 : &game->p1;

    fputs("+----+", out);
    for (i = 0; i < NUM_CUPS; i++)
        fprintf(out, "----+");
    fputs("----+\n", out);

    fputs("|    |", out);
    for (i = 0; i < NUM_CUPS; i++)
        fprintf(out, " %2hi |", far->cups[NUM_CUPS - i - 1]);
    fputs("    |\n", out);

    fprintf(out, "| %2hi |", far->mancala);
    for (i = 0; i < NUM_CUPS; i++)
        fprintf(out, "----+");
    fprintf(out, " %2hi |\n", near->mancala);

    fputs("|    |", out);
    for (i = 0; i < NUM_CUPS; i++)
        fprintf(out, " %2hi |", near->cups[i]);
    fputs("    |\n", out);

    fputs("+----+", out);
    for (i = 0; i < NUM_CUPS; i++)
        fprintf(out, "----+");
    fputs("----+\n", out);
}

File: mancala.py
import itertools
from typing import NamedTuple


class Mancala(NamedTuple):
    """
    Represents the state of a mancala game.
    """

    cups: tuple[tuple[int, int], ...]
    ends: tuple[int, int]
    p1_turn: bool

    def __str__(self) -> str:
        n = len(self.cups)
        top = "│".join(f" {c:2} " for _, c in self.cups)
        bot = "│".join(f" {c:2} " for c, _ in self.cups)

        return "\n".join(
            [
                "┌────┬────" + "┬────" * (n - 1) + "┬────┐",
                f"│ {self.ends[1]:2} │" + top + "│    │",
                "│    ├────" + "┼────" * (n - 1) + "┤    │",
                "│    │" + bot + f"│ {self.ends[0]:2} │",
                "└────┴────" + "┴────" * (n - 1) + "┴────┘",
            ]
        )

    @classmethod
    def start(cls) -> "Mancala":
        """
        Returns the starting state for a Mancala game, which is 6 pairs of cups
        with 4 stones in each.
        """
        return cls(tuple((4, 4) for _ in range(6)), (0, 0), True)

    def _near(self):
        """
        Returns a generator over the cups on the near side of the board, in the
        order from the first player's perspective.
        """
        return (c for c, _ in self.cups)

    def _far(self):
        """
        Returns a generator over the cups on the far side of the board, in the
        order from the second player's perspective.
        """
        return (c for _, c in reversed(self.cups))

    def game_over(self) -> bool:
        """
        A game is over if either player has no stones in any of their cups.
        """
        return all(c <= 0 for c, _ in self.cups) or all(c <= 0 for _, c in self.cups)

    def move(self, i_pick: int) -> "Mancala | None":
        """
        Performs a move for the player whose turn it is. `i_pick` is the index
        of the cup on the player's side to pick from and shall be in the range
        [0, n).

        If the selected cup is empty, then that move is not allowed an `None` is
        returned.
        """
        n = len(self.cups)
        x, _ = self.cups[i_pick]

        if x > 0:
            # distribute the stones around the board in a cycle
            q, r = divmod(x, 2 * n + 1)
            cycle = [
                q
                + (m if i != i_pick else 0)
                + (1 if (i - (i_pick + 1)) % (2 * n + 1) < r else 0)
                for i, m in enumerate(
                    itertools.chain(self._near(), (self.ends[0],), self._far())
                )
            ]

            near = cycle[:n]
            far_reversed = cycle[:n:-1]
            e0 = cycle[n]
            e1 = self.ends[1]

            # opposite cup stealing
            i_last = (i_pick + r) % (2 * n + 1)
            if i_last < n and near[i_last] == 1 and far_reversed[i_last] > 0:
                e0 += far_reversed[i_last] + 1
                near[i_last] = far_reversed[i_last] = 0

            # player goes again if last stone dropped in mancala
            return (
                Mancala(
                    cups=tuple(zip(near, far_reversed)),
                    ends=(e0, e1),
                    p1_turn=self.p1_turn,
                )
                if i_last == n
                else Mancala(
                    cups=tuple(zip(reversed(far_reversed), reversed(near))),
                    ends=(e1, e0),
                    p1_turn=not self.p1_turn,
                )
            )
